### PR TITLE
Update pull requests based on the result of the pipelinerun

### PR DESCRIPTION
**Pull request by akihikokuroda**
# Changes
This is the refresh PR of https://github.com/tektoncd/experimental/pull/154

When the webhook triggers due to a pull request on the git repo, a pipelinerun is created along with a task to monitor its status.  This task updates the pull request with the pipeline run result.
The default task used to update the pull request is monitor-result-task and the comment added to the pull request is either "OK" or "ERROR" along with a message based on the result of the pipelinerun.

The monitoring task can be overridden, on creation of the webhook (using the REST endpoint only) by setting the `pulltask` property in the POST body. The associated comments that get added to the pull request based on success or failure of the pipelinerun can be set using the `onsuccesscomment` (for success) and `onfailurecoment` (for failure) properties on the REST request.

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [ ] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [ ] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)

_See [the contribution guide](https://github.com/tektoncd/experimental/blob/master/CONTRIBUTING.md)
for more details._

### CODE REVIEW

File: webhooks-extension/DEVELOPMENT.md [hunk 1]
```diff
@@ -46,10 +46,20 @@ data='{
...
   "pipeline": "simple-pipeline",
   "dockerregistry": "mydockerregistry"
+  "pulltask": "monitor-result-task",
+  "onsuccesscomment": "OK",
+  "onfailurecomment": "ERROR",
 }'
 curl -d "${data}" -H "Content-Type: application/json" -X POST http://localhost:8080/webhooks
 ```
 
+`pulltask`: task name that monitors the pipeline and updates the pull request when it finishes. The dfault task is `monitor-result-task`
```
[the remark this answers is not in the document]
Comment (akihikokuroda) [reply webhooks-extension/DEVELOPMENT.md]: fixed the typo.   Add one statement about the mandatory items.

File: webhooks-extension/DEVELOPMENT.md [hunk 2]
```diff
@@ -46,10 +46,20 @@ data='{
...
+
+`onsuccesscomment`: comment text that is put into the pull request when the pipelinerun finishes sccessfully.  The default is `OK: <pipelinerun name>`
```
[the remark this answers is not in the document]
Comment (akihikokuroda) [reply webhooks-extension/DEVELOPMENT.md]: fixed

File: webhooks-extension/config/release/gcr-tekton-webhooks-extension.yaml [hunk 3]
```diff
@@ -146,4 +146,76 @@ spec:
         env:
         - name: INSTALLED_NAMESPACE
           value: tekton-pipelines
+        - name: SERVICEACCOUNT
+          value: tekton-webhooks-extension
```
[the remark this answers is not in the document]
Comment (akihikokuroda) [reply webhooks-extension/config/release/gcr-tekton-webhooks-extension.yaml]: No, the knative service doesn't take it.

File: webhooks-extension/config/release/gcr-tekton-webhooks-extension.yaml [hunk 4]
```diff
@@ -146,4 +146,76 @@ spec:
...
+  inputs:
+    resources:
+      - name: pull-request
+        type: pullRequest
+    params:
+      - name: pipelinerun
+        description: The name of the pipelinerun to be monitored
+        default: default pipelinerun
+      - name: commentsuccess
+        description: The text of the sccess comment
```
[the remark this answers is not in the document]
Comment (akihikokuroda) [reply webhooks-extension/config/release/gcr-tekton-webhooks-extension.yaml]: fixed

File: webhooks-extension/config/release/gcr-tekton-webhooks-extension.yaml [hunk 5]
```diff
@@ -146,4 +146,76 @@ spec:
...
+        default: default success at task
```
[the remark this answers is not in the document]
Comment (akihikokuroda) [reply webhooks-extension/config/release/gcr-tekton-webhooks-extension.yaml]: Change the default text.
It doesn't change the state of the PR.

File: webhooks-extension/config/task-monitor-result.yaml [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+      cat <<EOF | python
+      import time, os, json
+      from kubernetes import client, config
+      config.load_incluster_config()
+      api_instance = client.CustomObjectsApi(client.ApiClient(client.Configuration()))
+      comment = "$COMMENT_FAILURE"
+      i = range(100)
+      for x in i:
+          time.sleep( 10 )
+          output = api_instance.get_namespaced_custom_object("tekton.dev", "v1alpha1", "$NAMESPACE", "pipelineruns", "$PIPELINERUN")
```
[the remark this answers is not in the document]
Comment (akihikokuroda) [reply webhooks-extension/config/task-monitor-result.yaml]: I'm not sure.  It should be investigated later.

File: webhooks-extension/pkg/endpoints/sink.go [hunk 7]
```diff
@@ -99,12 +100,15 @@ func (r Resource) handleWebhook(request *restful.Request, response *restful.Resp
...
+		buildInformation.PULLURL = strings.Replace(webhookData.PullRequest.URL, "api/v3/repos/", "", 1)
 		buildInformation.TIMESTAMP = timestamp
 		buildInformation.BRANCH = extractBranchFromRef(webhookData.PullRequest.Head.Ref)
 
-		createPipelineRunFromWebhookData(buildInformation, r)
+		pipelinerun := createPipelineRunFromWebhookData(buildInformation, r)
 		logging.Log.Debugf("Build information for repository %s:%s: %s.", buildInformation.REPOURL, buildInformation.SHORTID, buildInformation)
-
+		createTaskRunFromWebhookData(buildInformation, r, pipelinerun)
+		logging.Log.Debugf("created monitoring task for pipelinerun from build information for repository %s sha %s.", buildInformation.REPOURL,
```
[the remark this answers is not in the document]
Comment (akihikokuroda) [reply webhooks-extension/pkg/endpoints/sink.go]: changed.

File: webhooks-extension/pkg/endpoints/sink.go [hunk 8]
```diff
@@ -233,16 +237,110 @@ func createPipelineRunFromWebhookData(buildInformation BuildInformation, r Resou
...
+	return pipelineRun
+}
+
+// This creates TaskRun for monitoring the main PipelineRun and reporting the result to the github
+func createTaskRunFromWebhookData(buildInformation BuildInformation, r Resource, pipelinerun *v1alpha1.PipelineRun) {
+	logging.Log.Debugf("In createTaskRunFromWebhookData, build information: %s.", buildInformation)
+
+	installNs := os.Getenv("INSTALLED_NAMESPACE")
+	if installNs == "" {
+		installNs = "default"
```
[the remark this answers is not in the document]
Comment (akihikokuroda) [reply webhooks-extension/pkg/endpoints/sink.go]: Yes, it has been used as the default namespace in other place like the webhook configmap.  I change them.

File: webhooks-extension/pkg/endpoints/sink.go [hunk 9]
```diff
@@ -233,16 +237,110 @@ func createPipelineRunFromWebhookData(buildInformation BuildInformation, r Resou
...
+
+	// get information from related githubsource instance
+	webhook, err := r.getGitHubWebhook(buildInformation.REPOURL, installNs)
+	if err != nil {
+		logging.Log.Errorf("error getting github webhook: %s.", err.Error())
+		return
+	}
+
+	taskTemplateName := webhook.PullTask
+	taskNs := webhook.Namespace
```
[the remark this answers is not in the document]
Comment (akihikokuroda) [reply webhooks-extension/pkg/endpoints/sink.go]: Yes, I change the variable name.

**Comment by akihikokuroda** [comment]
Ready for another review.

File: webhooks-extension/DEVELOPMENT.md [hunk 10]
```diff
@@ -46,10 +46,21 @@ data='{
...
 curl -d "${data}" -H "Content-Type: application/json" -X POST http://localhost:8080/webhooks
 ```
 
+`pulltask`: task name that monitors the pipeline and updates the pull request when it finishes. The default task is `monitor-result-task`
+
+`onsuccesscomment`: comment text that is put into the pull request when the pipelinerun finishes successfully.  The default is `OK: <pipelinerun name>`
+
+`onfailurecomment`: comment text that is put into the pull request when the pipelinerun finishes with errors.  The default is `ERROR: <pipelinerun name>`
+
+The rest of items are mandatory in the POST data.
```
Comment (dibbles) [inline webhooks-extension/DEVELOPMENT.md]: I'm not sure I'm a fan of this ...... I would delete this and have a section before describing these 3 settings (i.e line 55) that states: "Whilst most parameters are mandatory the following three are optional.  Not setting them will allow the defaults to take effect and this is recommended in most cases."

Comment (akihikokuroda) [reply webhooks-extension/DEVELOPMENT.md]: OK.  I changed.

**Pull request closed without merging**